File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/RegisterRequest.java
package hottopic.mit.co.nz.cleaningservice.network;

import android.content.Context;

import java.util.Map;

import hottopic.mit.co.nz.cleaningservice.entities.network.BooleanResponse;
import hottopic.mit.co.nz.cleaningservice.entities.users.UserInfo;
import hottopic.mit.co.nz.cleaningservice.network.service.UserService;
import io.reactivex.Single;

/**
 * The request of registration.
 */
public class RegisterRequest extends BaseRequest<UserService, BooleanResponse, BooleanResponse> {
    private UserInfo userInfo;

    public RegisterRequest(Context context, UserInfo userInfo) {
        super(context);
        this.userInfo = userInfo;
    }

    @Override
    protected Map<String, String> getHeader() {
        return null;
    }

    @Override
    protected Class getEndpointClass() {
        return UserService.class;
    }

    @Override
    protected Single<BooleanResponse> getEndpoint(UserService endpoint) {
        return endpoint.register(userInfo);
    }

    @Override
    protected BooleanResponse dealWithResponse(BooleanResponse response) {
        return response;
    }
}

File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/RegisterRequest.kt
package hottopic.mit.co.nz.cleaningservice.network

import android.content.Context

import hottopic.mit.co.nz.cleaningservice.entities.network.response.BooleanResponse
import hottopic.mit.co.nz.cleaningservice.entities.users.UserInfo
import hottopic.mit.co.nz.cleaningservice.network.service.UserService
import io.reactivex.Single

/**
 * The request of registration.
 */
class RegisterRequest(context: Context, private val userInfo: UserInfo) : BaseRequest<UserService, BooleanResponse, BooleanResponse>(context) {

    override val header: Map<String, String>?
        get() = null

    override val endpointClass: Class<UserService>
        get() = UserService::class.java

    override fun getEndpoint(endpoint: UserService?): Single<BooleanResponse?>? {
        return endpoint?.register(userInfo)
    }

    override fun dealWithResponse(response: BooleanResponse): BooleanResponse? {
        return response
    }
}